geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileReader.scala

```diff
@@ -58,127 +51,186 @@ class SimpleFeatureArrowFileReader(is: InputStream, filter: Filter = Filter.INCL
+      val feature = vector.reader.feature
+      val all = Iterator.range(0, total).map { i => vector.reader.load(i); feature }
       if (filter == Filter.INCLUDE) { all } else {
         all.filter(filter.evaluate)
       }
     }
+  }
+
+  /**
+    * Reads features from a simple feature vector. The underlying features are assumed to be sorted
+    *
+    * @param vector simple feature vector
+    * @param filter filter
+    * @param skip will be toggled if no further vectors need to be queried due to sort order and filter bounds
+    * @param bounds bounds for the sort field, extracted from the filter
+    * @param sortField field that the features are sorted by
+    * @param reverse if the sort order is reversed or not
+    * @return
+    */
+  private def sortedFeatures(vector: SimpleFeatureVector,
+                             filter: Filter,
+                             skip: SkipIndicator,
+                             bounds: Seq[Bounds[Comparable[Any]]],
+                             sortField: Int,
+                             reverse: Boolean): Iterator[ArrowSimpleFeature] = {
+    val total = vector.reader.getValueCount
+
+    if (total == 0 || skip.skip) { Iterator.empty } else {
+      // re-use the same feature object
+      val feature = vector.reader.feature
+
+      // bounds for the current batch
+      val b = {
+        vector.reader.load(0)
+        val lo = Option(feature.getAttribute(sortField).asInstanceOf[Comparable[Any]])
+        vector.reader.load(total - 1)
+        val hi = Option(feature.getAttribute(sortField).asInstanceOf[Comparable[Any]])
+        if (reverse) { Bounds(hi, lo, inclusive = true) } else { Bounds(lo, hi, inclusive = true) }
+      }
 
-    override def close(): Unit = {
-      reader.close()
-      vector.close()
+      if (bounds.forall(Bounds.intersection(_, b).isEmpty)) {
+        // nothing from this batch matches, check to see if any further batches could match
+        val hasMore = if (reverse) {
+          bounds.exists { bound =>
+            (bound.lower, b.lower) match {
+              case (Some(b1), Some(b2)) if b1.compareTo(b2) > 0 => false
+              case _ => true
+            }
+          }
+        } else {
+          bounds.exists { bound =>
+            (bound.upper, b.upper) match {
+              case (Some(b1), Some(b2)) if b1.compareTo(b2) < 0 => false
+              case _ => true
+            }
+          }
+        }
+        // toggle the skip indicator if there are no further batches that could match
+        skip.skip = !hasMore
+        Iterator.empty
+      } else {
+        val all = Iterator.range(0, vector.reader.getValueCount).map { i => vector.reader.load(i); feature }
+        all.filter(filter.evaluate)
+      }
     }
   }
-}
\ No newline at end of file
+
+  // holder for a skip indicator - this will be toggled if we ever determine there can be no more results
+  private [io] class SkipIndicator(var skip: Boolean = false)
```

Comment:
no, in this case we want to pass around the outer class, while allowing the inner skip flag to be toggled. it's not very 'functional', but with all the stacked arrow iterators it was a lot simpler to be able to flip this flag way down in the stack and have it available everywhere else.